lib/eventHubReceiver.ts

```diff
@@ -0,0 +1,382 @@
+    this.identifier = options.identifier;
+    this.options = options;
+    this.receiverRuntimeMetricEnabled = options.enableReceiverRuntimeMetric || false;
+    this.runtimeInfo = {
+      paritionId: `${partitionId}`
+    };
+
+    const onMessage = (context: rheaPromise.Context) => {
+      const evData = EventData.fromAmqpMessage(context.message);
+      this.emit(Constants.message, evData);
+    };
+    const onError = async (context: rheaPromise.Context) => {
+      // Since the receiver received an error the link has been closed. So calling
+      // this.close() will ensure that the receiver has been removed from the context.
+      await this.close();
+      this.emit(Constants.error, errors.translate(context.receiver.error));
+    };
+
+    this.on("newListener", (event) => {
+      if (event === Constants.message) {
+        if (this._session && this._receiver) {
+          debug("Attaching an event handler for the 'message' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.on(Constants.message, onMessage);
+        }
+      }
+
+      if (event === Constants.error) {
+        if (this._session && this._receiver) {
+          debug("Attaching an event handler for the 'receiver_error' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.on(Constants.receiverError, onError);
+        }
+      }
+    });
+
+    this.on("removeListener", (event) => {
+      if (event === Constants.message) {
+        if (this._session && this._receiver) {
+          debug("Removing an event handler for the 'message' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.removeListener(Constants.message, onMessage);
+        }
+      }
+
+      if (event === Constants.error) {
+        if (this._session && this._receiver) {
+          debug("Removing an event handler for the 'receiver_error' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.removeListener(Constants.receiverError, onError);
+        }
+      }
+    });
+  }
+
+  /**
+   * Creates a new AMQP receiver under a new AMQP session.
+   * @returns {Promoise<void>}
+   */
+  async init(): Promise<void> {
+    try {
+      // Acquire the lock and establish a cbs session if it does not exist on the connection.
+      await defaultLock.acquire(this._context.cbsSession.cbsLock, () => { return this._context.cbsSession.init(this._context.connection); });
+      const tokenObject = await this._context.tokenProvider.getToken(this.audience);
+      debug(`[${this._context.connectionId}] EH Receiver: calling negotiateClaim for audience "${this.audience}"`);
+      // Negotitate the CBS claim.
+      await this._context.cbsSession.negotiateClaim(this.audience, this._context.connection, tokenObject);
+      if (!this._session && !this._receiver) {
+        let receiverError: any;
+        let rcvrOptions: rheaPromise.ReceiverOptions = {
+          autoaccept: true,
```

Comment:
Yes that is the same in rhea. I did not want to manage credits for an Event hubs receiver since we want to receive as many messages possible and that too at a faster rate. Hence autoaccept: true works out well. But in Servicebus I can totally see autoaccept being set to false for certain scenarios (release/modified, etc.).

Reply:
what do other SDKs do and should this be part of the receiver options in case a user want to benefit/use this amqp feature?

---
In reply to: [180937198](https://github.com/Azure/azure-event-hubs-node/pull/47#discussion_r180937198) [](ancestors = 180937198)